src/Castle.ts

```diff
@@ -0,0 +1,218 @@
+import fetch from 'node-fetch';
+import { reduce } from 'lodash';
+import { IncomingHttpHeaders } from 'http2';
+import AbortController from 'abort-controller';
+
+const defaultApiUrl = 'https://api.castle.io';
+
+type TrackParameters = {
+  event: string;
+  user_id: string;
+  user_traits: any;
+  created_at?: string;
+  context: {
+    ip: string;
+    client_id: string;
+    headers: IncomingHttpHeaders;
+  };
+};
+
+type AuthenticateResult = {
+  action: string;
+  user_id?: string;
+  device_token?: string;
+  failover?: boolean;
+  failover_reason?: string;
+};
+
+const isTimeout = (e: Error) => e.name === 'AbortError';
+
+export class Castle {
+  private apiSecret: string;
+  private apiUrl: string;
+  private timeout: number;
+  private allowedHeaders: string[];
+  private disallowedHeaders: string[];
+  private overrideFetch: any;
+  private failoverStrategy: string;
+
+  constructor({
+    apiSecret,
+    apiUrl,
+    timeout = 500,
+    allowedHeaders = [],
+    disallowedHeaders = [],
+    overrideFetch = fetch,
+    failoverStrategy = 'allow',
+  }: {
+    apiSecret: string;
+    apiUrl?: string;
+    timeout?: number;
+    allowedHeaders?: string[];
+    disallowedHeaders?: string[];
+    overrideFetch?: any;
+    failoverStrategy?: string;
+  }) {
+    if (!apiSecret) {
+      throw new Error(
+        'Castle: Unable to instantiate Castle client, API secret is missing.'
+      );
+    }
+
+    this.apiSecret = apiSecret;
+    this.apiUrl = apiUrl || defaultApiUrl;
+    this.timeout = timeout;
+    this.allowedHeaders = allowedHeaders;
+    this.disallowedHeaders = disallowedHeaders;
+    this.overrideFetch = overrideFetch;
+    this.failoverStrategy = failoverStrategy;
+  }
+
+  public async authenticate(
+    params: TrackParameters
+  ): Promise<AuthenticateResult> {
+    if (!params.event) {
+      throw new Error('Castle: event is required when calling authenticate.');
```

Comment:
could we extract all the error messages somewhere to have all the potential errors listed in one place?

Reply:
Sure

Reply:
should we have similar structure and behaviour like in other sdks - it would be easier for anyone to do a change across all of them

Reply:
Event is a very minimal requirement to actually run. Doesn't castle-ruby check for event?

I think @wallin is working on a SDK specification document, then we can make sure all SDKs follow it.

Reply:
I mean it does but there is an additional file structure for that like /validators /commands etc.. I dont say that all is good in ruby sdk but it would be great to keep file structure somehow in sync (not exact but in some level similar across sdks) it would be much easier to add new features to sdks.

Reply:
and easy to compare if we miss somewhere or not as well